Getting Funqy Ready for Release (#9344)
quarkusio/quarkus: docs/src/main/asciidoc/funqy-knative-events.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Funqy Knative Events

include::./attributes.adoc[]
:extension-status: preview

Quarkus Funqy link:https://knative.dev/docs/eventing[Knative Events] builds off of the link:funqy-http[Funqy HTTP] extension to allow you to
route and process Knative Events within a Funqy function.

The guide walks through quickstart code to show you how you can deploy and invoke on Funqy functions
with Knative Events.

== Prerequisites

To complete this guide, you need:

* 60+ minutes
* Read about link:funqy[Funqy Basics].  This is a short read!
* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
* Apache Maven {maven-version}
* Have gone through the link:https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/index.html[Knative Tutorial], specifically link:https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial-eventing/eventing-trigger-broker.html[Brokers and Triggers]

== Setting up Knative

> Setting up Knative locally in a Minikube environment is beyond the scope of this guide.  You must
> walk through this link:https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/index.html[Knative Tutorial]

Reviewer: ```suggestion
Setting up Knative locally in a Minikube environment is beyond the scope of this guide.  It is advised
to follow  https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/index.html[this] Knative Tutorial
```

link:funqy-http[Funqy HTTP] using normal HTTP requests, or Cloud Event Binary mode, or Structured Mode.  All
invocation modes are supported at the same time.

So, if you open up the unit test code in {quickstarts-tree-url}/funqy-quickstarts/funqy-knative-events-quickstart/src/test/java/org/acme/funqy/FunqyTest.java[FunqyTest.java]
you'll see that its simply using RestAssured to make HTTP invocations to test the functions.

Funqy also works with Quarkus Dev mode!

== Build the Project

First build the Java artifacts:

[source, shell]
----
mvn clean install
----

Next, a docker image is required by Knative, so you'll need to build that next:

[source, shell]
----
docker build -f src/main/docker/Dockerfile.jvm -t yourAccountName/funqy-knative-events-quickstart .
----

Make sure to replace `yourAccountName` with your docker or quay account name when you run `docker build`.  The
Dockerfile is a standard Quarkus dockerfile.  No special Knative magic.

Push your image to docker hub or quay

[source, shell]
----
docker push yourAccountName/funqy-knative-events-quickstart
----

Again, make sure to replace `yourAccountName` with your docker or quay account name when you run `docker push`.

> == Deploy to Kubernetes/Openshift

Reviewer: In a subsequent revisions, we should see how we can integrate this with the Quarkus Kubernetes extensions that has some support for KNative
